pkg/epp/scheduling/plugins/prefix/plugin.go

```diff
@@ -0,0 +1,178 @@
+	// Why not just return the server with longest prefix match?
+	// It may not be the optimal choice, e.g., it may have a high queue depth.
+	// We optimistically search more than one to give more candidates for the scheduler to choose.
+	DefaultNumServersToMatch = 2
+	// vLLM default token block size is 16, and a good guess of average characters per token is 4.
+	DefaultCacheBlockSize  = 64
+	DefaultMaxPrefixBlocks = 128
+	// Assume each request reaches DefaultMaxPrefixBlocks = 128, and each BlockHash is cached onto 2
+	// servers due to load balancing, then it requires 256 entries per request.
+	// According to the estimates in indexer.estimateEntrySize(), the size of each entry is 348 bytes.
+	// So each request will cost 89,088 bytes ~ 90KB.
+	// Therefore, to cache 50k requests, we need 50K * 90KB = 4.5GB. Assuming 500 requests per
+	// second, a 4.5 GB cache can hold at least last 100 seconds of requests.
+	// Note in practice, the size of each entry will be much smaller (shorter NamespacedNames,
+	// shorter prompt). And due to the prefix cache hit, the number of unique cache entries will be
+	// much smaller per request. Therefore the actual cache size will be much smaller.
+	// TODO: Add some guidance for choosing the right size.
+	DefaultLRUIndexerCapacity = 50000
+)
+
+type Config struct {
+	// The input prompt is broken into sizes of HashBlockSize to calculate block hashes . Requests
+	// with length shorter than the block size will be ignored.
+	HashBlockSize int
+	// MaxPrefixBlocksToMatch is the maximum number of prefix blocks to match. Input beyond this limit will
+	// be ignored.
+	MaxPrefixBlocksToMatch int
+	// Max (approximate) size of the LRU indexer in number of entries.
+	LRUIndexerCapacity int
+}
+
+var DefaultConfig = Config{
+	HashBlockSize:          DefaultCacheBlockSize,
+	MaxPrefixBlocksToMatch: DefaultMaxPrefixBlocks,
+	LRUIndexerCapacity:     DefaultLRUIndexerCapacity,
+}
+
+type plugin struct {
+	Config
+	indexer Indexer
+}
+
+type Indexer interface {
+	Get(hash types.BlockHash) map[types.ServerID]bool
+	Add(hashes []types.BlockHash, server types.ServerID)
+}
+
+func New(config Config) *plugin {
+	m := &plugin{
+		Config:  config,
+		indexer: newIndexer(config.LRUIndexerCapacity),
+	}
+	return m
+}
+
+func (m *plugin) Name() string {
+	return "prefixCache"
+}
+
+func (m *plugin) PreSchedule(ctx *types.SchedulingContext) {
+	ctx.PrefixHashes = hashPrompt(ctx, m.HashBlockSize, m.MaxPrefixBlocksToMatch)
+	ctx.PrefixCacheServers = m.matchLongestPrefix(ctx, DefaultNumServersToMatch)
+	ctx.Logger.V(logutil.DEBUG).Info(fmt.Sprintf("PreSchedule, cached servers: %+v", ctx.PrefixCacheServers), "hashes", ctx.PrefixHashes)
+}
+
+// If a request was routed to a server, record it in the cache:
+func (m *plugin) PostSchedule(ctx *types.SchedulingContext, res *types.Result) {
+	targetPod := res.TargetPod.GetPod()
+	m.indexer.Add(ctx.PrefixHashes, types.ServerID(targetPod.NamespacedName))
```

Comment:
this code assumes that request was sent successfully? PostSchedule is after taged pod was selected but before the request is sent to that pod. the request might fail sending to the target pod due to various reasons and there might be changes (like using a fallback pod). I think this should be post response instead of post schedule with the actual pod that got the request.

Reply:
In practice I don't how much difference it is, given the likely hood of request failing is low. Also using PostSchedule also has the advantage of solving the "head of line blocking" when EPP just starts or a burst of requests, since it doesn't need to wait for response.

Reply:
Even if we guessed wrong (e.g., request failed or a fallback endpoint is picked), the cost is that next request is sent to a server which may not have the cache. However, after that the prefix should be cached and following requests should still get the cache hit.